src/main/java/com/snowflake/kafka/connector/internal/StageFilesProcessor.java

```diff
@@ -0,0 +1,622 @@
+  @FunctionalInterface
+  interface TaskAwaiter {
+    void await(long duration) throws InterruptedException;
+  }
+
+  public StageFilesProcessor(
+      String pipeName,
+      String tableName,
+      String stageName,
+      String prefix,
+      SnowflakeConnectionService conn,
+      SnowflakeIngestionService ingestionService,
+      SnowflakeTelemetryPipeStatus pipeTelemetry,
+      SnowflakeTelemetryService telemetryService) {
+    this(
+        pipeName,
+        tableName,
+        stageName,
+        prefix,
+        conn,
+        ingestionService,
+        pipeTelemetry,
+        telemetryService,
+        System::currentTimeMillis);
+  }
+
+  @VisibleForTesting
+  StageFilesProcessor(
+      String pipeName,
+      String tableName,
+      String stageName,
+      String prefix,
+      SnowflakeConnectionService conn,
+      SnowflakeIngestionService ingestionService,
+      SnowflakeTelemetryPipeStatus pipeTelemetry,
+      SnowflakeTelemetryService telemetryService,
+      TimeSupplier currentTimeSupplier) {
+    this.pipeName = pipeName;
+    this.tableName = tableName;
+    this.stageName = stageName;
+    this.prefix = prefix;
+    this.conn = conn;
+    this.currentTimeSupplier = currentTimeSupplier;
+    this.ingestionService = ingestionService;
+    this.telemetryService = telemetryService;
+    this.cleanerExecutor = ForkJoinPool.commonPool();
+    this.pipeTelemetry = pipeTelemetry;
+    this.filters = new FilteringPredicates(currentTimeSupplier, prefix);
+  }
+
+  /**
+   * Starts stage files processor in the background. Uses one of the common thread pool's thread.
+   * The processing keeps running until close() is invoked. Processing is executed every minute.
+   *
+   * @return instance of ProgressRegister - client interface to interact with the worker thread in a
+   *     thread safe manner.
+   */
+  public ProgressRegister trackFilesAsync() {
+    SnowflakeTelemetryPipeCreation pipeCreation =
+        new SnowflakeTelemetryPipeCreation(tableName, stageName, pipeName);
+    ProgressRegisterImpl register = new ProgressRegisterImpl(this);
+
+    ProgressRegistryTelemetry telemetry =
+        new ProgressRegistryTelemetry(pipeCreation, pipeTelemetry, telemetryService);
+
+    cleanerExecutor.submit(
+        () ->
+            trackFiles(
+                register, telemetry, () -> !Thread.currentThread().isInterrupted(), Thread::sleep));
+
+    register.waitForStart();
+    return register;
+  }
+
+  /**
+   * trackFiles - actual processor logic with extracted thread management for testing purposes.
+   *
+   * @param register - client interface
+   * @param progressTelemetry - telemetry interface
+   * @param canContinue - delegate to Thread.currentThread().isInterrupted(), required for testing
+   * @param awaiter - delegate to Thread.sleep(), required for testing
+   */
+  @VisibleForTesting
+  void trackFiles(
+      ProgressRegisterImpl register,
```

Comment:
should we use `ProgressRegister` interface here?

Reply:
implementation has some internal content - thus - the implementation here. the trackFiles itself should be considered private (shall we put it in dedicated package? - it would be effecitvely private for other callers)

Reply:
it will make testing difficult. but I dont have a major concern here.